src/uploadAssets.js

```diff
@@ -214,24 +214,32 @@ function getUploadFailures(assetRefMap, assetIds) {
 function setAssetReferences(assetRefMap, assetIds, options) {
   const {client, tag} = options
   const lookup = assetRefMap.values()
-  const patchTasks = assetIds.reduce((tasks, assetId) => {
+
+  // Collects patch tasks per document to avoid patching the same document multiple times
+  const patchTasksPerDoc = assetIds.reduce((tasks, assetId) => {
     const documents = lookup.next().value
     if (typeof assetId !== 'string') {
       return tasks
     }
 
-    return tasks.concat(
-      documents.map(({documentId, path}) => ({
-        documentId,
-        path,
-        assetId,
-      })),
-    )
-  }, [])
+    documents.forEach(({documentId, path}) => {
+      tasks[documentId] = tasks[documentId] || []
+      tasks[documentId].push({path, assetId})
+    })
+    return tasks
+  }, {})
+
+  const patchTasks = []
+  Object.keys(patchTasksPerDoc).forEach((documentId) => {
+    patchTasks.push({
+      documentId,
+      tasks: patchTasksPerDoc[documentId],
+    })
+  })
```

Comment:
SImpler if written as a map operation?

```suggestion
  const patchTasks = Object.entries(patchTasksPerDoc).
      map(([documentId, tasks]) => ({documentId, tasks}))
```